src/nvim/normal.c

```diff
@@ -2096,17 +2096,23 @@ static void display_showcmd(void)
   grid_line_flush();
 }
 
+int get_vtopline(win_T *wp)
+{
+  return plines_m_win_fill(wp, 1, wp->w_topline) - wp->w_topfill;
+}
+
 /// When "check" is false, prepare for commands that scroll the window.
 /// When "check" is true, take care of scroll-binding after the window has
 /// scrolled.  Called from normal_cmd() and edit().
 void do_check_scrollbind(bool check)
 {
   static win_T *old_curwin = NULL;
-  static linenr_T old_topline = 0;
-  static int old_topfill = 0;
+  static linenr_T old_vtopline = 0;
   static buf_T *old_buf = NULL;
   static colnr_T old_leftcol = 0;
 
+  int vtopline = get_vtopline(curwin);
```

Comment:
Doing this is more expensive than before. I wonder if there is a way to avoid doing this everytime.

Reply:
This should be efficient now as `plines_m_win_fill` now invokes just one iterator on the marktree.